reservation 사용자 예매 현황, 판매 현황 기능 추가

taesukang-dev opened the pull request:
## 📌Linked Issues
- #30 

## ✏Change Details
- 현황 조회와 관련된 기능을 추가하였습니다.
- show(공연) - show detail(회차) - seat(좌석) 의 3차원 배열의 구조 입니다.
- 각각의 항목을 grouping 하여 사용자가 확인하기 쉽게 하기 위함입니다.
- 판매 현황과 사용자의 예매 현황을 같은 response 구조로 사용하고 있습니다.
- 단, 사용자는 어떤 공연의 어떤 회차의 좌석을 몇 매 구매했는지 확인할 수 있게 했고,
- 판매자는 각 회차의 판매 현황을 확인해야 하기 때문에 모든 seat (총 좌석과, 잔여 좌석을 볼 수 있는 entity) unique 하게 가져왔습니다.

## 💬Comment
- 여러 엔티티를 projection 하면서 dao 라는 객체를 생성하여 한 데 담아주었는데, 옳은 방법인지 모르곘습니다.

## 📑References
- 없음

## ✅Check List
- [x] 추가한 기능에 대한 테스트는 모두 완료하셨나요?
- [x] 코드 정렬(Ctrl + Alt + L), 불필요한 코드나 오타는 없는지 확인하셨나요?

taesukang-dev committed:
chore: 공백 제거

taesukang-dev committed:
[#30] feat: 사용자 예매 조회 기능 추가

접속한 사용자의 email 을 기준으로 예매한 내역을 조회하는 기능을 추가하였습니다.
show - show detail - seat 를 단계로 ReservationResponse 를 기준으로 3차원 배열과 같은 구조로 작성하였습니다.
사용자가 조회할 때에는 같은 공연인지, 회차인지, 좌석 몇 매인지에 대한 구분을 쉽게 하기 위함입니다.

taesukang-dev committed:
[#30] style: response 명 변경

응답하는 변수명들을 각각 show, showdetails, seats 로 변경하였습니다.

taesukang-dev committed:
[#30] feat: 판매 현황 조회 기능 추가

사용자 예매 조회와 동일한 구조, 기능을 가진 판매 현황 조회 기능을 추가하였습니다.
seller dao 를 따로 두면서 사용자 예매 기능과 함께 사용할 수 있도록 개발하였습니다.

taesukang-dev committed:
[#30] fix: 판매 현황 기능과 함께 쓰기 위한 수정

사용자의 예매 현황, 판매자의 판매 현황을 같은 구조로 쓸 수 있다는 판단으로 개발하였습니다.
- dao 를 따로 쓰면서 reservation response 에서 ofseller 라는 메소드를 따로 만들어 reservation dao 로 casting 하였습니다.
- seat response 에서 current quantity, total quantity 를 추가하였습니다.
- 다만 판매 현황에서 seat response 의 count 는 무조건 1입니다.

taesukang-dev committed:
[#30] test: 판매 현황 service test 추가

판매 현황의 서비스 테스트 코드를 추가하였습니다.

sungyujeon commented:
제가 다른 대안이 따로 떠오르지 않아서 어떻게 개선될 수 있을지 궁금하네요..
구조가 결론적으로는 repository에서 쿼리하여 특정 사용자의 예매내역 또는 판매내역을 모두 들고온 다음에 Response 객체를 조립하기 위해 dto 객체가 많이 사용된 것으로 이해하였는데요.

1. join 쿼리 시 데이터베이스에서 User, Show, ShowDetail, Seat, Reservation의 모든 정보를 reservation 쿼리 결과 수만큼 join 해서 불필요하게 많은 결과를 반환시켜야 할 것으로 보이는데요. 그래서 response 객체들에서도 contains() 로직으로 조립 시 중복을 줄인 것으로 보였습니다. 이걸 어떻게 바꿀 수 있을지 모르겠네요..

2. 각 도메인별 정보들을 중첩 리스트 형식으로 모두 조합하여 response 해주어야 할 때는 어떻게 처리하는 것이 좋을지 고민이 되네요..

f-lab-joseph reviewed (changes requested):
멘토링 시간 리뷰

taesukang-dev committed:
[#30] fix: 쿼리 및 dto 수정

entity 를 담는 대신 개별 컬럼으로 수정하였습니다.

taesukang-dev committed:
[#30] test: 현황 조회 관련 테스트 코드 수정

reservation dto fixture 객체 추가 및 
seller 현황 조회 스펙 변경에 따른 test code 수정

taesukang-dev committed:
[#30] fix: seller 현황 조회 수정

기존 3차원 배열 구조에서 flat 하게 변경하였습니다.
그에 따라 seat, showdetail response 객체를 삭제하였습니다.

taesukang-dev committed:
[#30] fix: 예매 현황 조회 로직 수정

기존 3차원 구조에서 예매 매수(seat id) 기준으로 grouping 하여 반환하도록 하였습니다.
예매를 생성하면서 개별로 save 쿼리를 실행하던 것을 save all 로 변경하였습니다.

taesukang-dev commented:
> 멘토링 시간 리뷰

@f-lab-joseph 
@sungyujeon 

https://github.com/f-lab-edu/IMHERO/pull/34/commits/08a11d381fa389b2d26e9c2583410d8dff1f38cb
https://github.com/f-lab-edu/IMHERO/pull/34/commits/fb475d4050e6258d7c8b207afc18300837f874f9
https://github.com/f-lab-edu/IMHERO/pull/34/commits/85b66fca7d221d911562ef33345c72d9b10fec9d
https://github.com/f-lab-edu/IMHERO/pull/34/commits/af9f470528c1c86c9bd37b2f61fed0772eabccba

위의 4 커밋을 통해 수정하였습니다.
1. reservation - save 메서드에서 개별로 save 하던 로직을 saveAll 로 변경하였습니다.
2. 판매자의 현황 조회를 flat 하게 변경하였습니다.
3. 사용자의 예매 현황을 좌석 기준으로 (공연-회차-예매 매수) 정보를 담은 element 를 list 로 반환하였습니다.
